### PR TITLE
Return the output of the optimizer step

### LINKED ISSUE #5235: Lighting wrapped optimizers do not allow return values
This is both a bug report /feature request:

Any optimizer.step function might have a return value. However, NO return values are currently handled by the optimizer.LightningOptimizer class in the "step" function. Optimizers with return values  in their step function therefore do not operate properly.

BugFix: Add return value to to the respective line, something like:
```
return_val=None
if make_optimizer_step:
      return_val=self.__optimizer_step(*args, closure=closure, profiler_name=profiler_name, **kwargs)
....

return return_val
```
Also the `__optimizer_step` function must be able to pass on the return value

cc @borda @rohitgr7

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #5235

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @rohitgr7

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1546,18 +1546,19 @@ def optimizer_step(
...
+        This method (and ``zero_grad()``) won't be called during the accumulation phase when
+        ``Trainer(accumulate_grad_batches != 1)``. Overriding this hook has no benefit with manual optimization.
 
         Args:
             epoch: Current epoch
             batch_idx: Index of current batch
             optimizer: A PyTorch optimizer
             optimizer_idx: If you used multiple optimizers, this indexes into that list.
-            optimizer_closure: Closure for all optimizers. This closure must be executed as it includes the
+            optimizer_closure: The optimizer closure. This closure must be executed as it includes the
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:1561]: can we add a small example down below representing the value returned from optimizer.step()?

Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:1561]: slightly offtopic, but important IMO. Can you add a note that passing the closure into the optimizer.step() is required to make sure training_step, backward... is called? I have seen a few issues over slack/forums regarding that.

File: pytorch_lightning/lite/wrappers.py [hunk 2]
```diff
@@ -56,7 +56,7 @@ def state_dict(self) -> Dict[str, Tensor]:
 
     def step(self, closure: Optional[Callable] = None) -> None:
```
Comment (rohitgr7) [inline pytorch_lightning/lite/wrappers.py:57]:
```suggestion
    def step(self, closure: Optional[Callable] = None) -> Any:
```

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/core/lightning.py:1561]: The examples below and `optimizer_closure` description already mention this.

Technically passing it into the step is not required if you don't use `LBFGS`, what's required is that it's always executed.

Comment (carmocca) [reply pytorch_lightning/core/lightning.py:1561]: I don't think we need to because the `LightningOptimizer` is invisible to the user. After this PR it works exactly as if it did not exist. So it's more on PyTorch to describe any uses for this behavior.

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 3]
```diff
@@ -90,8 +90,10 @@ def optimizer_step(
         # in manual optimization, the closure does not return a value
```
Comment (akihironitta) [inline pytorch_lightning/plugins/precision/native_amp.py:90]: If I understand it correctly, this comment is no longer true.
```suggestion
```

**Review by akihironitta — APPROVED** [review]

File: pytorch_lightning/plugins/precision/native_amp.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/plugins/precision/native_amp.py:90]: It is true. Why do you think it isnt?

**Review by ananthsub — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

File: tests/core/test_lightning_optimizer.py [hunk 4]
```diff
@@ -314,3 +317,31 @@ def test_lightning_optimizer_keeps_hooks(tmpdir):
...
     del lightning_optimizer
     assert len(optimizer._fwd_handles) == 1
+
+
+def test_params_groups_and_state_are_accessible(tmpdir):
+    class TestModel(BoringModel):
+        def training_step(self, batch, batch_idx, optimizer_idx):
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+            self.__loss = loss
```
Comment (tchaton) [inline tests/core/test_lightning_optimizer.py:327]:
```suggestion
            self._loss = loss
```
